File: cpp/interview/classic/thread_safe_queue/thread_safe_queue.cpp
// 线程安全有界阻塞队列：mutex + 两个 condition_variable
#include <bits/stdc++.h>
#include <cassert>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <atomic>
using namespace std;

template <class T>
class BlockingQueue {
public:
    explicit BlockingQueue(size_t cap) : cap_(cap) {
        if (cap == 0) throw invalid_argument("cap must be positive");
    }

    void push(T item) {
        unique_lock<mutex> lk(mtx_);
        not_full_.wait(lk, [&] { return closed_ || buf_.size() < cap_; });
        if (closed_) throw runtime_error("queue closed");
        buf_.push_back(std::move(item));
        not_empty_.notify_one();
    }

    T pop() {
        unique_lock<mutex> lk(mtx_);
        not_empty_.wait(lk, [&] { return closed_ || !buf_.empty(); });
        if (buf_.empty()) throw runtime_error("queue closed");
        T v = std::move(buf_.front());
        buf_.pop_front();
        not_full_.notify_one();
        return v;
    }

    bool tryPop(T& out, chrono::milliseconds timeout) {
        unique_lock<mutex> lk(mtx_);
        if (!not_empty_.wait_for(lk, timeout, [&] { return closed_ || !buf_.empty(); })) {
            return false;
        }
        if (buf_.empty()) return false;
        out = std::move(buf_.front());
        buf_.pop_front();
        not_full_.notify_one();
        return true;
    }

    void close() {
        {
            lock_guard<mutex> lk(mtx_);
            closed_ = true;
        }
        not_full_.notify_all();
        not_empty_.notify_all();
    }

    size_t size() const {
        lock_guard<mutex> lk(mtx_);
        return buf_.size();
    }

private:
    size_t cap_;
    deque<T> buf_;
    mutable mutex mtx_;
    condition_variable not_full_;
    condition_variable not_empty_;
    bool closed_ = false;
};

int main() {
    BlockingQueue<int> q(8);
    const int nProducer = 4;
    const int per = 1000;
    const int total = nProducer * per;

    vector<thread> producers;
    for (int i = 0; i < nProducer; ++i) {
        producers.emplace_back([&q, i] {
            for (int j = 0; j < per; ++j) q.push(i * per + j);
        });
    }

    atomic<bool> stop{false};
    mutex outMtx;
    vector<int> consumed;
    consumed.reserve(total);

    vector<thread> consumers;
    for (int i = 0; i < 4; ++i) {
        consumers.emplace_back([&] {
            int v;
            while (true) {
                if (q.tryPop(v, chrono::milliseconds(50))) {
                    lock_guard<mutex> lk(outMtx);
                    consumed.push_back(v);
                } else if (stop.load() && q.size() == 0) {
                    return;
                }
            }
        });
    }

    for (auto& t : producers) t.join();
    stop.store(true);
    for (auto& t : consumers) t.join();

    assert((int)consumed.size() == total);
    sort(consumed.begin(), consumed.end());
    consumed.erase(unique(consumed.begin(), consumed.end()), consumed.end());
    assert((int)consumed.size() == total);

    cout << "thread_safe_queue OK" << endl;
    return 0;
}

File: python/interview/classic/thread_safe_queue/thread_safe_queue.py
"""手写线程安全有界阻塞队列。"""

from __future__ import annotations

import threading
from collections import deque
from typing import Generic, Optional, TypeVar

T = TypeVar("T")


class Closed(Exception):
    pass


class BlockingQueue(Generic[T]):
    def __init__(self, capacity: int) -> None:
        if capacity <= 0:
            raise ValueError("capacity must be positive")
        self._cap = capacity
        self._buf: deque[T] = deque()
        self._lock = threading.Lock()
        self._not_full = threading.Condition(self._lock)
        self._not_empty = threading.Condition(self._lock)
        self._closed = False

    def put(self, item: T, timeout: Optional[float] = None) -> bool:
        with self._not_full:
            if not self._wait_for(self._not_full, lambda: len(self._buf) < self._cap, timeout):
                return False
            self._buf.append(item)
            self._not_empty.notify()
            return True

    def get(self, timeout: Optional[float] = None) -> T:
        with self._not_empty:
            if not self._wait_for(self._not_empty, lambda: len(self._buf) > 0, timeout):
                raise TimeoutError("get timed out")
            item = self._buf.popleft()
            self._not_full.notify()
            return item

    def close(self) -> None:
        with self._lock:
            self._closed = True
            self._not_full.notify_all()
            self._not_empty.notify_all()

    def qsize(self) -> int:
        with self._lock:
            return len(self._buf)

    def _wait_for(self, cond: threading.Condition, pred, timeout: Optional[float]) -> bool:
        ok = cond.wait_for(lambda: self._closed or pred(), timeout=timeout)
        if self._closed:
            raise Closed("queue closed")
        return ok


if __name__ == "__main__":
    q: BlockingQueue[int] = BlockingQueue(8)
    n_producer = 4
    per = 1000
    total = n_producer * per

    def producer(start: int) -> None:
        for i in range(per):
            q.put(start * per + i)

    consumed: list[int] = []
    consumed_lock = threading.Lock()

    stop = threading.Event()

    def consumer() -> None:
        while True:
            try:
                v = q.get(timeout=0.5)
            except TimeoutError:
                if stop.is_set() and q.qsize() == 0:
                    return
                continue
            with consumed_lock:
                consumed.append(v)

    ts = [threading.Thread(target=producer, args=(i,)) for i in range(n_producer)]
    cs = [threading.Thread(target=consumer) for _ in range(4)]
    for c in cs:
        c.start()
    for t in ts:
        t.start()
    for t in ts:
        t.join()
    stop.set()
    for c in cs:
        c.join()

    assert len(consumed) == total
    assert len(set(consumed)) == total
    print("thread_safe_queue OK")
